amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseCleanupManager.java

```diff
@@ -0,0 +1,338 @@
+        completedLeaseStopwatch.start();
+        garbageLeaseStopwatch.start();
+
+        deletionThreadPool.scheduleAtFixedRate(new LeaseCleanupThread(), INITIAL_DELAY, leaseCleanupIntervalMillis,
+                TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Enqueues a lease for deletion.
+     * @param leasePendingDeletion
+     */
+    public void enqueueForDeletion(LeasePendingDeletion leasePendingDeletion) {
+        final Lease lease = leasePendingDeletion.lease();
+        if (lease == null) {
+            log.warn("Cannot enqueue lease {} for deferred deletion - instance doesn't hold the lease for that shard.",
+                    lease.leaseKey());
+        } else {
+            //TODO: Optimize verifying duplicate entries https://sim.amazon.com/issues/KinesisLTR-597.
+            if (!deletionQueue.contains(leasePendingDeletion)) {
+                log.debug("Enqueuing lease {} for deferred deletion.", lease.leaseKey());
+                deletionQueue.add(leasePendingDeletion);
+            } else {
+                log.warn("Lease {} is already pending deletion, not enqueueing for deletion.", lease.leaseKey());
+            }
+        }
+    }
+
+    /**
+     * Returns how many leases are currently waiting in the queue pending deletion.
+     * @return number of leases pending deletion.
+     */
+    public int leasesPendingDeletion() {
+        return deletionQueue.size();
+    }
+
+    private boolean timeToCheckForCompletedShard() {
+        return completedLeaseStopwatch.elapsed(TimeUnit.MILLISECONDS) >= completedLeaseCleanupIntervalMillis;
+    }
+
+    private boolean timeToCheckForGarbageShard() {
+        return garbageLeaseStopwatch.elapsed(TimeUnit.MILLISECONDS) >= garbageLeaseCleanupIntervalMillis;
+    }
+
+    private LeaseCleanupResult cleanupLease(LeasePendingDeletion leasePendingDeletion) throws TimeoutException,
+            InterruptedException, DependencyException, ProvisionedThroughputException, InvalidStateException {
+        final Lease lease = leasePendingDeletion.lease();
+        final ShardInfo shardInfo = leasePendingDeletion.shardInfo();
+        final StreamIdentifier streamIdentifier = leasePendingDeletion.streamIdentifier();
+
+        final AWSExceptionManager exceptionManager = createExceptionManager();
+
+        boolean cleanedUpCompletedLease = false;
+        boolean cleanedUpGarbageLease = false;
+        boolean alreadyCheckedForGarbageCollection = false;
+
+        try {
+            if (cleanupLeasesUponShardCompletion && timeToCheckForCompletedShard()) {
+                Set<String> childShardKeys = leaseCoordinator.leaseRefresher().getLease(lease.leaseKey()).childShardIds();
+                if (CollectionUtils.isNullOrEmpty(childShardKeys)) {
+                    try {
+                        childShardKeys = getChildShardsFromService(shardInfo, streamIdentifier);
+                        updateLeaseWithChildShards(leasePendingDeletion, childShardKeys);
```

Comment:
can you check for childShardKeys not null and log error if otherwise?

Reply:
Will add